fe2-android/app/src/androidTest/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivityTest.java

```
@@ -0,0 +1,182 @@
+    homeButton().perform(click());
+
+    assertEquals(
+        HomeActivity.class.getName(),
+        activityScenarioRule
+            .getScenario()
+            .getResult()
+            .getResultData()
+            .getComponent()
+            .getClassName());
+  }
+
+  @Test
+  public void identityButtonOpensIdentityTab() {
+    identityButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(identityFragmentId()))));
+  }
+
+  @Test
+  public void showPropertyButtonShowsConnectQRCode() {
+    showPropertiesButton().perform(click());
+
+    propertiesLayout().check(matches(isDisplayed()));
+
+    String expectedQRCode = gson.toJson(new ConnectToLao(requestFactory.getUrl(), LAO_ID));
+    connectQrCode().check(matches(withQrCode(expectedQRCode)));
+  }
+
+  @Test
+  public void witnessButtonShowsWitnessTab() {
+    witnessButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(witnessFragmentId()))));
+  }
+
+  // Matches an ImageView containing a QRCode with expected content
+  private Matcher<? super View> withQrCode(String expectedContent) {
+    return new TypeSafeMatcher<View>() {
+      @Override
+      protected boolean matchesSafely(View item) {
+        if (item instanceof ImageView) {
+          ImageView image = (ImageView) item;
+          Drawable drawable = image.getDrawable();
+
+          if (drawable instanceof BitmapDrawable) {
+            Bitmap qrcode = ((BitmapDrawable) drawable).getBitmap();
+            // Convert the QRCode to something zxing understands
+            int[] buffer = new int[qrcode.getWidth() * qrcode.getHeight()];
+            qrcode.getPixels(
+                buffer, 0, qrcode.getWidth(), 0, 0, qrcode.getWidth(), qrcode.getHeight());
+            LuminanceSource source =
+                new RGBLuminanceSource(qrcode.getWidth(), qrcode.getHeight(), buffer);
+            BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));
```

this looks like a good candidate for a separate conversion function.
you could also just throw an exception if the drawable is not of the right type.

---

Yes, it's better